gem/optimise.py

```diff
@@ -260,6 +286,76 @@ def count(pair):
     return product, flops
 
 
+def associate_sum(summands):
+    """Apply associativity rules to construct an operation-minimal summation tree.
+    """
+    # group summands by their free indices
+    groups = OrderedDict()
+    for summand in summands:
+        key = frozenset(summand.free_indices)
+        groups.setdefault(key, []).append(summand)
+    summands = [reduce(Sum, _value, Zero()) for _value in itervalues(groups)]
+
+    if len(summands) > 32:
+        # O(N^3) algorithm
+        raise NotImplementedError("Not expected such a complicated expression!")
+
+    def count(pair):
+        """Operation count to add a pair of GEM expressions"""
+        a, b = pair
+        extents = [i.extent for i in set().union(a.free_indices, b.free_indices)]
+        return numpy.prod(extents, dtype=int)
+
+    flops = 0
+    while len(summands) > 1:
+        # Greedy algorithm: choose a pair of factors that are the
+        # cheapest to add.
+        a, b = min(combinations(summands, 2), key=count)
+        flops += count((a, b))
+        # Remove chosen factors, append their product
+        summands.remove(a)
+        summands.remove(b)
+        summands.append(Sum(a, b))
+    sum, = summands
+    return sum, flops
```

Comment:
Okay, this is basically duplication of codes that already exist for products, this should be refactored to avoid this duplication.